fecgov/fecfile-web-api
# 651 - Add structured application logs

## Summary

- Issue api#651
  
Add structlog for structured logging. 

Note - intention was to use `json` logs in cloud.gov, but there's an issue over there (See https://github.com/fecgov/fecfile-web-api/issues/651#issuecomment-1922408723) so my recommendation is to move forward with better local logs and use key/value pairs in production until the json issue can be addressed by cloud.gov

## How to test

- Run locally to see new API logs, log in with the front end (http://localhost:4200/) and then go to http://localhost:8080/celery-test/ to see celery logs
- To test key-value settings for cloud.gov, make the below changes, restart celery and go to http://localhost:8080/celery-test/. Check out API and celery logs

**django-backend/fecfiler/settings/base.py**
```python
# LOG_FORMAT = env.get_credential("LOG_FORMAT", LINE)
LOG_FORMAT = KEY_VALUE
```

## Tasks
- [x] Rebase off develop branch
- [x] Manual deploy to dev space
- [X] Check wtih devs on logging output
- [x] Review logs in cloud.gov https://logs.fr.cloud.gov/goto/561b230039570eb4ae4043306e6001ac
- [x] Test a non-debug celery task (`debug_task` dumps in all the request context)
- [x] Review logs to look for opportunities for better data, clean up old-style string formatting to make key/value additions easier - start with automated tests
- [x] Look into exception handler
- [x] Add tests for logs! https://www.obeythetestinggoat.com/how-to-log-exceptions-to-stderr-in-django.html
- [x] New fields
- [x] Check on logging preferences for end-to-end tests
- [x] Remove deploy logic changes from tasks.py and circle config
- [x] Ask cloud.gov for a ticket and/or docs on where the json limitations stand (their public documentation could use some clarity as well): https://github.com/cloud-gov/logsearch-boshrelease/issues/161
